### ISSUE TITLE
Command registration scopes

### ISSUE THREAD

**Issue by matteodelabre** [sile-typesetter/sile#1510]
Would it make sense to add a scoping mechanism for command registration? Calling something like `registerScopedCommand("x", …)` during the processing of a parent command would temporarily define `\x` (potentially shadowing an existing definition of `\x`) until we leave the parent scope and `\x` reverts to its previous definition, if any.

The best application I could find is for `\item`, which currently only makes sense inside of an `\itemize` or `\enumerate` from the `lists` package. But the `\item` name is generic enough that it’s not hard to imagine use cases for it meaning different things based on the context. The shadowing mentioned above would be useful when nesting “environments” in which the `\item` command has different meanings.

Of course a solution to this is namespacing, but it’d be a bit cumbersome for authors to have to write `\lists:item` for each item.

Another case is with the `\mo`, `\mi` and `\mn` commands from `math`, which only make sense when inside of `\mathml`. I’m also working on a package to typeset algorithm pseudocode, where commands like `\if`, `\else`, …, only make sense inside of an algorithm.

Just trying to gather some feedback before I actually implement anything.

**Comment by Omikhleia** [comment]
@matteodelabre One possible approach is _not_ to have registered commands for all `\xxx` keywords, but to handle them in a toplevel command that looks at the AST.
- In the SILE manual, for an illustration of this, look at how the DocBook `<title>` is extracted in the "SILE As An XML Processor" section... It's about XML processing, but the mechanism is general, e.g. it's also how one can "extract" some sub-node and handle it separately (I use this in my own classes/packages, for example, to extract the `\caption` in captioned figures or tables -- see e.g. [here](https://github.com/Omikhleia/sile/blob/dbc6e556ac817f28fea9a871c8c47b1609bcfb3f/packages/markdown/commands.lua#L262) (Caveat: This link will probably be wrong when I rebase and update this stuff a some point)
- You may also have a look at my PR #1344 for a similar logic : the `\row`, `\cell` etc. only exist within a (p)table environment.
- In the case of the existing **lists** package (providing itemize/enumerate), I went for a slightly different approach as I felt having `\item` available outside those list is useful e.g. for warning the user that he's using it wrong -- but there's a bit of AST munching here too when being _inside_ a list.

That is to say, at the "low level", SILE can already achieve something like this. The user input is converted to an AST that is processed (via `SILE.process` etc.), but at each command level you may tweak how the AST (living in the `content` structure passed to commands) is handled. In other terms, whatever the input (SIL, XML, ...), it's converted to an AST that one manipulate programmatically.

A general utility of the kind you describe -- an abstraction of the above -- would thus be doable. It doesn't even have to be in the "core" parsing logic, and can reside in a package (that people can opt to use or not). I have actually been toying too with a similar concept, for XML processing again though (but really it doesn't make a true difference), with some helpers such as `registerStructuredCommand("tag", "allowed sub tags list"...)` to make my XML handling easier -- but I haven't yet proceeded far in that direction, having a bunch of other ToDos in my task list ;) -- But I find the idea sound, if it provides enough abstraction to be generalized.

To sum up:
- I do think it's doable ;)
- It can be a package providing some higher level registration and processing mechanism, without actually changing anything in the core parser (provided one sticks to the current AST, which I would recommend...).

The obvious _cons_ is that it can make life harder for users to override those commands, but there are cases where this is acceptable IMHO. (E.g. I can't think to any use case when "redefining" a table cell _really_ makes sense, esp. if the latter has some more accessible hook later).

**Comment by Omikhleia** [comment]
Obnote: as for "temporary" command redefinitions (shadowing a command, possibly keeping the ability to still use it, and restoring it), you can also find ideas [here](https://github.com/Omikhleia/omikhleia-sile-packages/blob/main/packages/redefine.lua) -- It's a SILE 0.12 package, I have a hard time considering when to upgrade all my packages to 0.14, but still you'd get an additional idea about another possible way... I have to say I very seldom used this package however (one of my very first packages, lol), except for quick hacks ;)

**Comment by Omikhleia** [comment]
This issue did not received any activity for several years and the discussions pointed towards other ways to achieve this.
I am closing it, but please feel free to re-open a dedicated issue if a different approach should be discussed.

**Comment by alerque** [comment]
I'm sorry I probably should have chipped in sooner, but I'm going to re-open because I think there is something here we need to consider. I've had my eye on it for a while.

Note we already *do* a lot of this, we just have to do it by jumping through hoops each and every time. Typically this is done by redefining the *parent command*, then stashing a copy of the existing command we want to change, subbing in a new definition, using it, then swapping it back out for the original that we stashed before relinquishing control of the parent command.

This works to the job done in most cases but is actually fairly dangerous for 3rd party packages to do as a regular pattern when when all they really want is to catch something in a specific scope.

**Comment by alerque** [comment]
Note this is also tangentially related to #1521. We're still quite away from that being a reality (although the mlua v0.10 release brought us one big step closer!) but whatever interface we do settle on for this will also interact with that.
